Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs

```diff
@@ -0,0 +1,153 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Abstractions;
+using System.Linq;
+
+namespace FileSystemVisitorLibrary
+{
+    public class FileSystemVisitor
+    {
+        private readonly Func<FileSystemInfo, bool> _filterFunc;
+
+        private readonly IFileSystem _fileSystem;
+
+        private Actions? _action = Actions.Continue;
+
+        public FileSystemVisitor(IFileSystem fileSystem, Func<FileSystemInfo, bool> filterFunc = null)
+        {
+            _fileSystem = fileSystem;
+            _filterFunc = filterFunc;
+        }
+
+        public event Action Started;
+
+        public event Action Complited;
+
+        public event Func<FileSystemInfo, Actions> FileFound;
+
+        public event Func<FileSystemInfo, Actions> DirectoryFound;
+
+        public event Func<FileSystemInfo, Actions> FilteredFileFound;
+
+        public event Func<FileSystemInfo, Actions> FilteredDirectoryFound;
+
+        public IEnumerable<string> Search(string path)
+        {
+            if (path == null)
+            {
+                throw new ArgumentNullException(path, "Directory path is null");
+            }
+
+            if (path == string.Empty)
+            {
+                throw new ArgumentException("Directory path is empty", path);
```

Comment:
Please, do not forget to use `nameof` keyword.

```suggestion
                throw new ArgumentException("Directory path is empty", nameof(path));
```